File: ios/Classes/Wrapper/Commands/TrackCustomEventCommand.swift
//
//  Created by Emarsys on 2021. 08. 11..
//
import EmarsysSDK

class TrackCustomEventCommand: EmarsysCommandProtocol {
    
    func execute(arguments: [String : Any]?, resultCallback: @escaping ResultCallback) {
        guard let eventName = arguments?["eventName"] as? String else {
            resultCallback(["error": "Invalid eventName"])
            return
        }
        var eventAttributes: [String: String]? = nil
        if let attributes = arguments?["eventAttributes"] as? [String: String] {
            eventAttributes = attributes
        }
        Emarsys.trackCustomEvent(eventName: eventName, eventAttributes: eventAttributes) { error in
            if let e = error {
                resultCallback(["error": e])
            } else {
                resultCallback(["success": true])
            }
        }
    }
}

File: android/src/main/kotlin/com/emarsys/emarsys_sdk/command/mobileengage/TrackCustomEventCommand.kt
package com.emarsys.emarsys_sdk.command.mobileengage

import com.emarsys.Emarsys
import com.emarsys.emarsys_sdk.command.EmarsysCommand
import com.emarsys.emarsys_sdk.command.ResultCallback

class TrackCustomEventCommand : EmarsysCommand {
    override fun execute(parameters: Map<String, Any?>?, resultCallback: ResultCallback) {
        val eventName: String? = parameters?.get("eventName") as String?
        eventName ?: throw IllegalArgumentException("eventName should not be null!")

        val eventAttributes: Map<String, String>? =
            parameters?.get("eventAttributes") as Map<String, String>?

        Emarsys.trackCustomEvent(eventName, eventAttributes) {
            resultCallback.invoke(null, it)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        return true
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }
}